codec.go

```diff
@@ -0,0 +1,173 @@
+package codec
+
+import (
+	"bufio"
+	"errors"
+	"io"
+	"math"
+
+	"zikichombo.org/sound"
+	"zikichombo.org/sound/ops"
+	"zikichombo.org/sound/sample"
+)
+
+// ErrUnknownCodec is an error representing
+var ErrUnknownCodec = errors.New("unknown codec")
+
+// ErrUnsupportedSampleCodec can be used by codec implementations
+// which receive a request for encoding or decoding with a sample codec
+// that is unsupported or doesn't make sense.
+var ErrUnsupportedSampleCodec = errors.New("unsupported sample codec")
+
+// AnySampleCodec is a sample.Codec which is used as a wildcard sample.Codec
+// in this interface.  Codecs which are not based on PCM data, such as aac/mp3/ogg/vorbis
+//
+var AnySampleCodec = sample.Codec(-1)
+
+// Codec represents a way of encoding and decoding sound.
+type Codec struct {
+	// Priority defines the preference order when more than one codec
+	// matches an encoding or decoding demand.  Filename extensions can trigger
+	// demands for encoding and decoding.  Magic triggers demands for decoding
+	// All demands may be further qualified by specifying a sample.Codec, with
+	//
+	Priority int
+
+	// Extensions lists the filename extensions which this codec claims to support.
+	// Examples are .wav, .ogg, .caf.
+	Extensions []string // Filename extensions
+
+	// Magic represents the first few bytes, such as fLaC.
+	Magic string // Magic represents the first few bytes often
+
+	// DefaultSampleCodec gives a default or preferred sample codec for the codec.
+	// Some codecs, such as perception based compressed codecs, don't really have
+	// a defined sample.Codec and should use AnySampleCodec for this field.
+	DefaultSampleCodec sample.Codec // DefaultSampleCodec
+
+	// Decoder tries to turn an io.ReadCloser into a sound.Source.  In the event
+	// the decoder does not use a defined sample.Codec, the second return
+	// value should be AnySampleCodec if the error return value is nil.
+	Decoder func(io.ReadCloser) (sound.Source, sample.Codec, error)
```

Comment:
Could you clarify what you mean with "In the event/ the decoder does not use a defined sample.Codec"? Should a Decoder return a specific `sample.Codec` if the entire sound file is of only one sample.Codec (e.g. 32-bit little endian)? And it should return `AnySampleCodec` if the sound file contains more than one sample codec, e.g. each FLAC frame has a different sample bit rate?